tools/evaluate-query-plan/main.go

```diff
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/gogo/protobuf/proto"
+	"github.com/grafana/dskit/flagext"
+
+	"github.com/grafana/mimir/pkg/frontend/v2/frontendv2pb"
+	"github.com/grafana/mimir/pkg/querier/querierpb"
+	"github.com/grafana/mimir/pkg/streamingpromql"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type app struct {
+	rootUrl      string
+	expr         string
+	tenantID     string
+	instantQuery bool
+
+	// Instant query:
+	time flagext.Time
+
+	// Range query:
+	start flagext.Time
+	end   flagext.Time
+	step  time.Duration
+
+	planner *streamingpromql.QueryPlanner
+	logger  log.Logger
+}
+
+func main() {
+	a := &app{
+		planner: streamingpromql.NewQueryPlanner(streamingpromql.NewTestEngineOpts()),
+		logger:  log.NewLogfmtLogger(os.Stderr),
+	}
+
+	if err := a.Run(); err != nil {
+		fmt.Printf("Application failed with an error: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func (a *app) ParseFlags() error {
+	flag.StringVar(&a.rootUrl, "root-url", "http://localhost:8005", "root URL of querier")
+	flag.StringVar(&a.expr, "expr", "", "expression to evaluate")
```

Comment:
Can you clarify if this is a PromQL expression or something else?

Reply:
Done in 6fbd45a.